storage: TestLeaseExtensionNotBlockedByRead failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/cff3e8870a9f2f84feb37415f223ce6922487504

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestLeaseExtensionNotBlockedByRead PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1086092&tab=buildLog

```
W190109 08:10:49.079712 72491 jobs/registry.go:316  [n1] unable to get node liveness: node not in the liveness table
I190109 08:10:49.265625 72838 sql/event_log.go:135  [n1,intExec=optInToDiagnosticsStatReporting] Event: "set_cluster_setting", target: 0, info: {SettingName:diagnostics.reporting.enabled Value:true User:root}
I190109 08:10:49.434302 72722 sql/event_log.go:135  [n1,intExec=set-setting] Event: "set_cluster_setting", target: 0, info: {SettingName:version Value:2.1-3 User:root}
I190109 08:10:49.514591 72826 sql/event_log.go:135  [n1,intExec=disableNetTrace] Event: "set_cluster_setting", target: 0, info: {SettingName:trace.debug.enable Value:false User:root}
I190109 08:10:49.664691 72883 sql/event_log.go:135  [n1,intExec=initializeClusterSecret] Event: "set_cluster_setting", target: 0, info: {SettingName:cluster.secret Value:405a8ef4-b450-4a9f-8af8-8c445c74aa57 User:root}
I190109 08:10:49.739835 72523 sql/event_log.go:135  [n1,intExec=create-default-db] Event: "create_database", target: 50, info: {DatabaseName:defaultdb Statement:CREATE DATABASE IF NOT EXISTS defaultdb User:root}
I190109 08:10:49.787711 72872 sql/event_log.go:135  [n1,intExec=create-default-db] Event: "create_database", target: 51, info: {DatabaseName:postgres Statement:CREATE DATABASE IF NOT EXISTS postgres User:root}
I190109 08:10:49.841825 72491 server/server.go:1620  [n1] done ensuring all necessary migrations have run
I190109 08:10:49.842036 72491 server/server.go:1623  [n1] serving sql connections
I190109 08:10:49.847441 72524 storage/replica_command.go:349  [n1,split,s1,r19/1:/{Table/23-Max}] initiating a split of this range at key /Table/24 [r20]
I190109 08:10:49.916380 72491 util/stop/stopper.go:536  quiescing; tasks left:
3      node.Node: batch
2      [async] kv.TxnCoordSender: heartbeat loop
1      [async] transport racer
1      [async] storage.split: processing replica
1      [async] closedts-rangefeed-subscriber
1      [async] auto-upgrade
W190109 08:10:49.917701 72737 storage/store.go:1646  [n1,s1,r5/1:/{System/tse-Table/System…}] unable to gossip on capacity change: node unavailable; try another peer
W190109 08:10:49.920282 72901 internal/client/txn.go:507  [n1] failure aborting transaction: node unavailable; try another peer; abort caused by: log-event: result is ambiguous (server shutdown)
W190109 08:10:49.920638 72901 server/node.go:942  [n1] node=1: unable to log node_join event: log-event: result is ambiguous (server shutdown)
E190109 08:10:49.922391 72802 server/status.go:919  [n1,status] node unavailable; try another peer
I190109 08:10:49.922689 72802 server/server_update.go:63  [n1] failed attempt to upgrade cluster version, error: rpc error: code = Internal desc = node unavailable; try another peer
I190109 08:10:49.923669 72491 util/stop/stopper.go:536  quiescing; tasks left:
1      [async] transport racer
1      [async] storage.split: processing replica
1      [async] closedts-rangefeed-subscriber
I190109 08:10:49.927881 72491 util/stop/stopper.go:536  quiescing; tasks left:
1      [async] transport racer
1      [async] storage.split: processing replica
W190109 08:10:49.928611 72524 internal/client/txn.go:507  [n1,split,s1,r19/1:/{Table/23-Max}] failure aborting transaction: node unavailable; try another peer; abort caused by: log-range-event: kv/txn_interceptor_heartbeat.go:404: node already quiescing
E190109 08:10:49.929306 72524 storage/queue.go:844  [n1,split,s1,r19/1:/{Table/23-Max}] unable to split [n1,s1,r19/1:/{Table/23-Max}] at key "/Table/24": split at key /Table/24 failed: log-range-event: kv/txn_interceptor_heartbeat.go:404: node already quiescing
I190109 08:10:49.929557 72491 util/stop/stopper.go:536  quiescing; tasks left:
1      [async] transport racer
I190109 08:10:49.950846 72668 kv/transport_race.go:113  transport race promotion: ran 12 iterations on up to 314 requests
--- FAIL: TestLeaseExtensionNotBlockedByRead (1.12s)
	client_replica_test.go:1244: result is ambiguous (unknown status for command without MaxLeaseIndex at refreshProposalsLocked time (refresh reason: reasonNewLeaderOrConfigChange))

```

--- a/pkg/server/status.go
+++ b/pkg/server/status.go
@@ -277,6 +277,9 @@ func (s *statusServer) Allocator(
 				func(desc roachpb.RangeDescriptor) (bool, error) {
 					rep, err := store.GetReplica(desc.RangeID)
 					if err != nil {
+						if _, skip := err.(*roachpb.RangeNotFoundError); skip {
+							return true, nil // continue
+						}
 						return true, err
 					}
 					if !rep.OwnsValidLease(store.Clock().Now()) {
@@ -1238,6 +1241,9 @@ func (s *statusServer) Ranges(
 			err := storage.IterateRangeDescriptors(ctx, store.Engine(),
 				func(desc roachpb.RangeDescriptor) (bool, error) {
 					rep, err := store.GetReplica(desc.RangeID)
+					if _, skip := err.(*roachpb.RangeNotFoundError); skip {
+						return true, nil // continue
+					}
 					if err != nil {
 						return true, err
 					}

--- a/pkg/storage/client_replica_test.go
+++ b/pkg/storage/client_replica_test.go
@@ -1224,11 +1224,6 @@ func TestLeaseExtensionNotBlockedByRead(t *testing.T) {
 			t.Fatalf("replica descriptor for key %s not found", rKey)
 		}
 
-		curLease, _, err := s.GetRangeLease(context.TODO(), key)
-		if err != nil {
-			t.Fatal(err)
-		}
-
 		leaseReq := roachpb.RequestLeaseRequest{
 			RequestHeader: roachpb.RequestHeader{
 				Key: key,
@@ -1238,10 +1233,30 @@ func TestLeaseExtensionNotBlockedByRead(t *testing.T) {
 				Expiration: s.Clock().Now().Add(time.Second.Nanoseconds(), 0).Clone(),
 				Replica:    replDesc,
 			},
-			PrevLease: curLease,
 		}
-		if _, pErr := client.SendWrapped(context.Background(), s.DB().NonTransactionalSender(), &leaseReq); pErr != nil {
-			t.Error(pErr) // NB: don't fatal or shutdown hangs
+
+		for {
+			curLease, _, err := s.GetRangeLease(context.TODO(), key)
+			if err != nil {
+				t.Fatal(err)
+			}
+			leaseReq.PrevLease = curLease
+
+			_, pErr := client.SendWrapped(context.Background(), s.DB().NonTransactionalSender(), &leaseReq)
+			if _, ok := pErr.GetDetail().(*roachpb.AmbiguousResultError); ok {
+				log.Infof(context.Background(), "retrying lease after %s", pErr)
+				continue
+			}
+			if _, ok := pErr.GetDetail().(*roachpb.LeaseRejectedError); ok {
+				// Lease rejected? Try again. The extension should work because
+				// extending is idempotent (assuming the PrevLease matches).
+				log.Infof(context.Background(), "retrying lease after %s", pErr)
+				continue
+			}
+			if pErr != nil {
+				t.Errorf("%T %s", pErr.GetDetail(), pErr) // NB: don't fatal or shutdown hangs
+			}
+			break
 		}
 		// Unblock the read.
 		readBlocked <- struct{}{}

